src/libraries/System.Text.Json/ref/System.Text.Json.cs

```diff
@@ -797,18 +802,38 @@ public sealed partial class JsonPropertyNameAttribute : System.Text.Json.Seriali
         public JsonPropertyNameAttribute(string name) { }
         public string Name { get { throw null; } }
     }
-    [System.AttributeUsageAttribute(System.AttributeTargets.Assembly, AllowMultiple=true)]
+    [System.AttributeUsageAttribute(System.AttributeTargets.Class, AllowMultiple=true)]
     public sealed partial class JsonSerializableAttribute : System.Text.Json.Serialization.JsonAttribute
     {
         public JsonSerializableAttribute(System.Type type) { }
         public string? TypeInfoPropertyName { get { throw null; } set { } }
+        public System.Text.Json.Serialization.JsonSourceGenerationMode GenerationMode { get { throw null; } set { } }
     }
     public abstract partial class JsonSerializerContext
     {
-        protected JsonSerializerContext(System.Text.Json.JsonSerializerOptions? options) { }
+        protected JsonSerializerContext(System.Text.Json.JsonSerializerOptions? instanceOptions, System.Text.Json.JsonSerializerOptions? defaultOptions) { }
```

Comment:
I find these 2 options a bit confusing in the API.

Reply:
I needed a way to fetch the default options assumed by a context instance so that we can check whether the run-time options align (and in turn, whether we can call generated serialization logic). Taking this instance in the ctor seemed like the most reliable way to make sure the options instance is available when we need it.

Reply:
Maybe we could name them something different? `instanceOptions` and `defaultOptions` don't really tell me what the difference is between them.

Another reason I don't like the names is that it violates this design guideline:

✔️ DO use the same name for constructor parameters and a property if the constructor parameters are used to simply set the property.

Another thought that I had was to instead have a virtual method on `JsonSerializerContext` that the generated code could override, and then the base `JsonSerializerContext` class wouldn't need to take the `defaultOptions`. But I can't think of a good name for that method, so I'm not certain I like that approach better. Something like `CanUseSerializationLogic`, `CanUseSerializationDelegates` or `HasDefaultOptions`.

I'm not sure my idea is better, but just listing it in case it sparks an idea for you.

Reply:
Thanks, I'll definitely explore these options as part of the next API review.